service/preinit.go

```diff
@@ -0,0 +1,170 @@
+// Copyright © 2024 Dell Inc. or its subsidiaries. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//      http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package service
+
+import (
+	"fmt"
+	"io/fs"
+	"os"
+	"strings"
+
+	"golang.org/x/net/context"
+)
+
+const (
+	defaultNodeMdmsFile = "/data/node_mdms.txt"
+)
+
+type PreInitService interface {
+	PreInit() error
+}
+
+func NewPreInitService() PreInitService {
+	return &service{}
+}
+
+type ArrayConfigurationProvider interface {
+	GetArrayConfiguration() ([]*ArrayConnectionData, error)
+}
+
+type FileWriterProvider interface {
+	WriteFile(filename string, data []byte, perm os.FileMode) error
+}
+
+type DefaultArrayConfigurationProvider struct{}
+
+func (s *DefaultArrayConfigurationProvider) GetArrayConfiguration() ([]*ArrayConnectionData, error) {
+	arrayConfig, err := getArrayConfig(nil)
+	if err != nil {
+		return nil, err
+	}
+
+	connectionData := make([]*ArrayConnectionData, 0)
+	for _, v := range arrayConfig {
+		connectionData = append(connectionData, v)
+	}
+
+	return connectionData, nil
+}
+
+type DefaultFileWriterProvider struct{}
+
+func (s *DefaultFileWriterProvider) WriteFile(filename string, data []byte, perm os.FileMode) error {
+	return os.WriteFile(filename, data, perm)
+}
+
+var (
+	arrayConfigurationProviderImpl ArrayConfigurationProvider = &DefaultArrayConfigurationProvider{}
+	fileWriterProviderImpl         FileWriterProvider         = &DefaultFileWriterProvider{}
+	nodeMdmsFile                                              = defaultNodeMdmsFile
+)
+
+func (s *service) PreInit() error {
+	Log.Infof("PreInit running")
+
+	arrayConfig, err := arrayConfigurationProviderImpl.GetArrayConfiguration()
+	if err != nil {
+		return fmt.Errorf("unable to get array configuration: %v", err)
+	}
+
+	labelKey, err := getLabelKey(arrayConfig)
+	if err != nil {
+		return fmt.Errorf("unable to get zone label key: %v", err)
+	}
+
+	var mdmData string
+
+	if labelKey == "" {
+		Log.Debug("No zone key found, will configure all MDMs")
+		sb := strings.Builder{}
+		for _, connectionData := range arrayConfig {
+			if connectionData.Mdm != "" {
+				if sb.Len() > 0 {
+					sb.WriteString("\\&")
+				}
+				sb.WriteString(connectionData.Mdm)
+			}
+		}
+		mdmData = sb.String()
+	} else {
+		Log.Infof("Zone key detected, will configure MDMs for this node, key: %s", labelKey)
+		nodeLabels, err := s.GetNodeLabels(context.Background())
```

Comment:
this method may need mild modification to be used here. 
GetNodeLabels uses this line:
https://github.com/dell/csi-powerflex/blob/9f67ab8994c740bec27ff8223dcd8389533af647/service/service.go#L1809
but I don't believe 
s.opts.KubeNodeName is set when running in init mode, I actually couldn't find where it was set at all in the driver. 
I think it would be good to make use of the env var NODENAME in GetNodeLabels so if s.opts.KubeNodeName = "", we can still grab node labels

Reply:
I was testing in driver and saw this error:
```
failed to complete pre-init: unable to get node labels: rpc error: code = Internal desc = Unable to fetch the node labels. Error: resource name may not be empty
```

Reply:
I made the following quick change where I build images to fix it:
```

diff --git a/service/service.go b/service/service.go
index 8c002e6..2dffe3c 100644
--- a/service/service.go
+++ b/service/service.go
@@ -1812,8 +1812,17 @@ func (s *service) GetNodeLabels(_ context.Context) (map[string]string, error) {
                K8sClientset = k8sutils.Clientset
        }

+       nodeName := s.opts.KubeNodeName
+
+       if  nodeName == "" {
+               Log.Infof("Using env variable for node name")
+               nodeName = os.Getenv("NODENAME")
+       }
+
+       Log.Infof("Using: %s as nodeName", nodeName)
+
        // access the API to fetch node object
-       node, err := K8sClientset.CoreV1().Nodes().Get(context.TODO(), s.opts.KubeNodeName, v1.GetOptions{})
+       node, err := K8sClientset.CoreV1().Nodes().Get(context.TODO(), nodeName, v1.GetOptions{})
```

I think we should do the same on this branch

Reply:
Added change and updated tests.